examples/pattern-matcher/define.scm

```diff
@@ -0,0 +1,63 @@
+;
+; define.scm
+;
+; Demonstrate the use of DefineLink to give names to things.
+;
+; XXX caution: under construction, unstable, known buggy.
+;
+(use-modules (opencog))
+(use-modules (opencog query))
+
+(load-from-path "utilities.scm")
+
+;; Some data to populate the atomspace.
+(InheritanceLink
+	(ConceptNode "battery")
+	(ConceptNode "electrical device"))
+
+(InheritanceLink
+	(ConceptNode "transistor")
+	(ConceptNode "electrical device"))
+
+(EvaluationLink
+	(PredicateNode "PartOf")
+	(ListLink
+		(ConceptNode "battery")
+		(VariableNode "car")))
+
+(EvaluationLink
+	(PredicateNode "PartOf")
+	(ListLink
+		(ConceptNode "transistor")
+		(VariableNode "phone")))
+
+(EvaluationLink
+	(PredicateNode "PartOf")
+	(ListLink
+		(ConceptNode "windsheild")
+		(VariableNode "car")))
+
+;; Define the concept of electrical parts of things.
+(define elect-parts
+	(DefineLink
+		(DefinedPredicateNode "Electrical Part Of")
+		(PresentLink
+			(InheritanceLink
+				(VariableNode "$x")
+				(ConceptNode "electrical device"))
+			(EvaluationLink
+				(PredicateNode "PartOf")
+				(ListLink
+					(VariableNode "$x")
+					(VariableNode "$y"))))))
+
+;; Define a pattern to find the electrical parts of things
+;; At this time, an explicit variable declaration has to be done;
+;; this needs to be fixed...
+(define get-elect
+	(GetLink 
+		(VariableList (VariableNode "$x") (VariableNode "$y"))
+		(DefinedPredicateNode "Electrical Part Of")))
```

Comment:
In your example "Electrical Part Of" is defined as a body with free
variables, then the variables are scoped by the GetLink. Wouldn't be
better if the variables where scoped in the DefineLink, such as

```
(define elect-parts
	(DefineLink
		(DefinedPredicateNode "Electrical Part Of")
        (LambdaLink
            (VariableList
                (VariableNode "$x")
                (VariableNode "$y"))
		    (PresentLink
                (InheritanceLink
				    (VariableNode "$x")
				    (ConceptNode "electrical device"))
			    (EvaluationLink
				    (PredicateNode "PartOf")
				    (ListLink
					    (VariableNode "$x")
					    (VariableNode "$y")))))))
```

then the GetLink can take the predicate alone

(GetLink (DefinedPredicateNode "Electrical Part Of"))

That way we're sure the variable names are consistent.